replica/tablets.cc

```diff
@@ -568,7 +568,7 @@ class tablet_incremental_selector : public sstables::incremental_selector_impl {
         auto token = pos.token();
         if (!_cur_set || pos.token() >= _lowest_next_token) {
             auto idx = _tset.group_of(token);
-            if (!token.is_maximum()) {
+            if (!token.is_maximum() && _tset._sstable_set_ids.contains(idx)) {
```

Comment:
I believe the incremental_selector should have been configured with only ranges that are contained within the replica. Even more, it should have been configured with a range that is fully contained within one tablet.

However, I don't want to block this series. Please file an issue about it. We'll need to then audit the various users who do full scans (storage_proxy::query_partition_key_range(), forward_service, repair, maybe more) and fix them up.

Reply:
we already intersect query ranges with tablet ranges using query_ranges_to_vnodes_generator (which impl depends on mode). that's true for forward_service::dispatch(), repair and storage_proxy::query_partition_key_range(). 

I might be wrong, but I think we only need _tset._sstable_set_ids.contains(idx) check for handling min_token on a full scan. (-inf, +inf), when intersected, will produce first range of (-inf, first_tablet_last_token], which is handled by first tablet of course. Maybe it can be fixed by changing sstables::incremental_reader_selector to treat -inf as first token (which is the first thing that can be consumed anyway). 

Maybe a patch as simple as this might do it:

```
diff --git a/sstables/sstable_set.cc b/sstables/sstable_set.cc
index 0df8b356b8..562bee932e 100644
--- a/sstables/sstable_set.cc
+++ b/sstables/sstable_set.cc
@@ -772,6 +772,7 @@ compaction_strategy::make_sstable_set(schema_ptr schema) const {
 using sstable_reader_factory_type = std::function<flat_mutation_reader_v2(shared_sstable&, const dht::partition_range& pr)>;
 
 static logging::logger irclogger("incremental_reader_selector");
+static const dht::token first_token = dht::first_token();
 
 // Incremental selector implementation for combined_mutation_reader that
 // selects readers on-demand as the read progresses through the token
@@ -795,7 +796,7 @@ class incremental_reader_selector : public reader_selector {
             const dht::partition_range& pr,
             tracing::trace_state_ptr trace_state,
             sstable_reader_factory_type fn)
-        : reader_selector(s, pr.start() ? pr.start()->value() : dht::ring_position_view::min())
+        : reader_selector(s, pr.start() ? pr.start()->value() : dht::ring_position_view::starting_at(first_token)
```

interestingly, later we might review forward_service::execute_on_this_shard(), which discards ranges not owned by this shard, but note that with sharder interface (IIRC, Avi has plans for changing it in the future), shard info only is not enough to detect that (as we need host id too), so range read can be executed by a shard that doesn't really own the range (but shares same shard id), so we should fail the request. ignoring the lack of presence of a sstable set prevents us from catching it.

Reply:
The intersection at the caller should recognize that (-inf, first_tablet_start_token) is empty and skip it, and generate ranges that are fully contained within a tablet.

Reply:
@Deexie could you please open a issue so we can keep track of it?

Reply:
https://github.com/scylladb/scylladb/issues/18684